### DISCUSSION TITLE
Must haves

### DISCUSSION THREAD

**Discussion by jhpratt** [rust-marker/design#1] [Ideas]
Use this discussion for suggestions of "must haves" for a custom linter. Please limit suggestions to one per top-level comment to allow for easier upvoting/downvoting and replies.

**Comment by jhpratt** [comment]
Lint implementations must be sandboxed.

**Reply by jhpratt** [reply]
NB: This is automatic with webassembly.

**Reply by flip1995** [reply]
Not sure if this is a must have. When using a dependency, you're supposed to do a code review before including it in your crate. At least, if you want to be save. The same can be expected before executing lints of a foreign crate.

**Reply by jhpratt** [reply]
If this becomes as widespread as I hope, doing a code review every time you want to even upgrade the patch version of a lint implementation seems infeasible. I know I wouldn't do it. By requiring sandboxing, it's not necessary to perform the code review, as the worst that can happen is the program hangs or emits crappy lints.

**Reply by flip1995** [reply]
Sandboxing the lint implementations seems like a good goal, but I wouldn't make it a "Must" goal. Must goals are goal that have to go in the MVP. This may hold us back.

**Reply by chorman0773** [reply]
This is really the same issue as proc-macros and build scripts. Those aren't sandboxed, and I wouldn't expect that lints need to be. There is also the question of whether or not it *should* be using wasm.

**Reply by jhpratt** [reply]
Build scripts can be used to generate files, so that's not quite the same. Personally I wish proc macros _were_ sandboxed for security reasons.

Regarding MVP, it shouldn't hold us back if wasm is used, as we'd get it for free. I think the benefits of wasm (from both a security and portability perspective) outweigh the drawbacks (shipping an interpreter or similar).

**Reply by flip1995** [reply]
Build scripts can execute arbitrary code.

I'm not convinced about the wasm approach either. Especially why we should force it.

**Reply by jhpratt** [reply]
Unless we're going to require the end user to compile the lint themselves, _someone_ has to build the lint for a number of different host targets. imo portability is sufficient reason for wasm, though security is also a huge plus.

Worth noting that both build scripts and proc macros exist to alter compilation in some way, while lints only exist to potentially halt it and give the user more info — no part of compilation actually depends on lints (excluding rustc lints for obvious reasons).

**Reply by chorman0773** [reply]
> Personally I wish proc macros were sandboxed for security reasons.

There were discussions about making those run in wasm as well. I agree with allowing that (and I think it is allowed), but I would raise the same concerns about forcing it. I really have no need to write all of the machinery necessary for using wasm inside the compiler, especially given I'd have to write that in C++ :grimacing:.

**Reply by chorman0773** [reply]
Couldn't lints in this world just be normal dependencies that either get run by default or on a special command line flag? I for one would like to have inline support for lints. Having to run an entirely separate binary to get lints is kind of annoying from clippy, especially when I typically forget to run all of those.

**Reply by jhpratt** [reply]
Having it be in a single binary would require compiler integration, no?

> I really have no need to write all of the machinery necessary for using wasm inside the compiler, especially given I'd have to write that in C++ 

Crates to compile and run wasm already exist, though? It's actually quite easy to use.

**Reply by chorman0773** [reply]
> Having it be in a single binary would require compiler integration, no?

If you mean having it attached to the compiler by default, then yes, yes it would. Implementations could use the same machinery for proc-macros as for lints. If you have cargo, then cargo handles the details. If you don't... Well, dealing with proc-macros in rust-autotools are annoying, but I see no issues with lints fundamentally, it's just another proc-macro.mk.in.

> Crates to compile and run wasm already exist, though? It's actually quite easy to use.

Sure, in rust. lccc, my implementation, is written in C++, and does not have a hard dependency on any host rust compiler (so that it can be bootstrapped), so if this was integrated into the base compiler, then I'd have to write the lint support directly into the compiler, which means that I'd have to write the wasm interpreter.

**Reply by jhpratt** [reply]
At least personally, I don't see the need to integrate this into the compiler, even long term. That was discussed in another comment chain, so I won't repeat it here. So long as it's standalone, alternative compiler implementations wouldn't need an interpreter.

**Reply by jhpratt** [reply]
Would this be acceptable to be placed under "could have"? I'm convinced this isn't necessary for the MVP, but I still think this is a good idea in the long term.

**Reply by jhpratt** [reply]
^^ Bueller?

**Comment by jhpratt** [comment]
The linter must be able to run in CI.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
The linter must provide a way to change the severity of the lint.

**Reply by flip1995** [reply]
This formulation is a bit confusing. I'm also not sure if we're on the same page with nomenclature:

Linter =  The consumer of the API implementing the lints

I think there are 2 separate things in this requirement:

1. The linter must be able to set the lint level of its lints
2. The crate the linter is run on must be able to change the lint levels

**Reply by jhpratt** [reply]
For clarity, this comment is the equivalent of `#[allow(foo)]` for existing lints. I think it's essential something like that exists here as well. It would be the linter (what we're trying to build) that actually provides that functionality, while the end user is the one who would invoke it. The lint implementation would have a default, which is mentioned elsewhere.

**Reply by flip1995** [reply]
I don't think we're trying to build a linter here. But this could just be different interpretations of the word "linter". A linter for me is the tool implementing lints. We won't implement any lints, but provide infrastructure that makes it easy to implement lints.

**Reply by jhpratt** [reply]
I'll open an issue regarding terminology.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
The lint author and the lint user must not need to know the implementation details of anything, including the linter, rustc crates, and the standard library.

**Reply by flip1995** [reply]
This is a "should" statement. Also do you mean rustc internals instead of standard library?

**Reply by jhpratt** [reply]
Difficult to get wording right! And yes, I meant rustc & friends. Although stdlib would also apply I suppose.

**Reply by flip1995** [reply]
The lint author may want to have knowledge of the stdlib, if the lints they implement depend on functions/structs/... available in std.

**Reply by jhpratt** [reply]
implementation details = stuff not visible through public APIs, no?

**Reply by chorman0773** [reply]
Indeed. Unless a lint author is specifically checking for code that relies on implementation-details, they shouldn't need to distinguish between standard library implementations

**Reply by chorman0773** [reply]
Though it should be possible to access those implementation-details, for that exact purpose.
If I'm writing a compatibility lint that is checking for code that is relying on an implementation-detail that changes between implementations, it's probably useful to know what that detail is.

**Reply by jhpratt** [reply]
Must not need to know ≠ cannot know. You could certainly create a lint to catch relying on implementation details, you just don't _have to_ know any implementation details for the general case.

**Reply by flip1995** [reply]
Ah I think the problem is the negative formulation. This is especially hard to understand for non-native English speakers (like me).

I would formulate it like this:

1. The lint author must be able to write lints, without knowing the implementation details of anything, including the ~~linter~~ stable API, rustc crates, or the standard library.
2. The lint users must be able to use the lints, without knowing the implementation details of the linter.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items. I used the wording from @flip1995.

**Comment by jhpratt** [comment]
All lints must work with the sole lint driver provided.

**Reply by chorman0773** [reply]
I think this can (and should) be debated. A good alternative would be, as mentioned in the sandboxing thread, to make lints just regular (or possibly dev) dependencies, and leave the details to a combination of the compiler and build system to handle.

**Reply by DevinR528** [reply]
Is one of the goals to move all linting back into the compiler? Don't we want a cargo install-able tool? I feel like the lints from the new API should work with the driver we design to work with them?

> to make lints just regular (or possibly dev) dependencies

I think this might be my one strong opinion, I don't love the idea of adding dependencies... although if the were just dev-deps and could be added by grouping so `pedantic = 1.0` to add all the pedantic lints that could work maybe?

**Reply by chorman0773** [reply]
> although if the were just dev-deps and could be added by grouping so pedantic = 1.0 to add all the pedantic lints that could work maybe?

Well, it would be adding something like `clippy = 1.5.2` or w/e. or `compat={path="../linters/compat"}`. One thing this does is keeps you from forgetting it. And I forget to run clippy alot (in the projects I haven't abandoned using it entirely). 

> Is one of the goals to move all linting back into the compiler? Don't we want a cargo install-able tool?

It could be. However, one  disadvantage of using 1 driver is that it's fixed to whatever implementation that was used to build it. It can't, for example, reach into the implementation-details of a different compiler, to check information that needs implementation knowledge (for example - abi).

**Reply by flip1995** [reply]
I think the best way to do this would be to add a section to the `Cargo.toml` file, where you can register linters / custom lints.

So something like:

```toml
[linters]
serde_lints = "1.0.53"
regex_lints = "0.15.2"
...
```

And this will then run the lints implemented by `serde` and `regex` and so on. Just like `cargo clippy` you may have to run `cargo lint` (or something similar), which will run all those extra lints. 

I don't think we should hook those lints directly into `cargo check` / `cargo build` / ... Running extra cargo commands is fine IMO. Forgetting this can't happen if you have a good CI setup and go through a PR workflow.

**Reply by chorman0773** [reply]
> I don't think we should hook those lints directly into cargo check / cargo build / ... Running extra cargo commands is fine IMO. Forgetting this can't happen if you have a good CI setup and go through a PR workflow.

Sure, if the CI has everything. However, I can imagine some code that would work fine on CI, but fail lints on my computer because some detected feature (yes, I typically use feature detection rather than relying on hardcoded behaviour for specific platforms) isn't available locally or on CI, or if CI doesn't have platform support for w/e

**Reply by flip1995** [reply]
>  I can imagine some code that would work fine on CI, but fail lints on my computer

CI should be the ground truth for your compilation, not your local machine. This is mainly because everyone has access to your CI, but only you have access to your machine. If it can pass your CI because it misses errors/warnings, than your CI setup is insufficient. If you have multiple features, you can run clippy with `--all-features`; multiple targets `--all-targets`. If this is not possible because enabling all features/targets at once breaks the code, you can run clippy multiple times in CI on different feature/target sets.

If you're CI can't support a specific feature or target for whatever reason, I would be careful to officially this target/feature, since you have no way to automatically test it.

---

What I want to say with all this: The argument, that you could forget to run a command isn't a valid concern IMHO. You could also forget to run `cargo build` before committing.

**Reply by chorman0773** [reply]
> If you're CI can't support a specific feature or target for whatever reason, I would be careful to officially this target/feature, since you have no way to automatically test it.

So, if CI supports some feature (meaning C function, `target_feature`, library, some function, etc.) I shouldn't support the fallback case? By feature detection I mean I'm doing autoconf-like tests to see what works until something does. It's harder to test when those feature tests fail if they would otherwise succeed. It also comes down to that CI is (or can be) limited. I can't exhaustively test all 167 targets supported by rustc. I can't really test however many Tier 2 or higher targets there are. I definately can't test every gcc-rs target, and I can't test every single quirk an end system could have that is tested by the build script. 

> What I want to say with all this: The argument, that you could forget to run a command isn't a valid concern IMHO. You could also forget to run cargo build before committing.

No, but having 1 binary *for each* lint set gets complex so that it's easier to forget at least one. Having one command in total to run all lints is fine, but I don't want to be running `cargo check; cargo clippy; cargo portability_check; cargo abi_check; cargo project-specific-lints` because I know I will forget one, and CI may or may not complain.

**Reply by flip1995** [reply]
If I would write software that claims to officially supports a specific target/feature, than I would want to test this in CI. If your requirement is to support every possible target+feature combination from every compiler there is, then yes you should test that on CI. But no one expects you to do this and this is not necessary in most cases. E.g. Clippy CI checks exactly 4 targets. It doesn't check for all targets out there.

> No, but having 1 binary for each lint set gets complex so that it's easier to forget at least one.

I agree. I would bundle this in one `cargo <command>` (which will be a cargo command we provide in addition to the stable API) that executes all the registered lints and later write an RFC that includes a `cargo lint` command in `cargo` that runs all the lints (+Clippy).

So the best case in the future would be that you run `cargo check` to get the rustc output and `cargo lint` to get the output of all the linters that are registered, including rustc lints (and Clippy lints). So practically instead of running `cargo check` you'll just run `cargo lint`. But this is far in the future.

**Reply by chorman0773** [reply]
Yeah that would be good. And if we are treating lints just as another class of dependency, we don't need to worry about the one driver thing, as $RUSTC and the build system can all handle that.

**Reply by jhpratt** [reply]
When I say "lint driver", I'm basically saying don't repeat what dylint does. Having a stable wrapper API would be sufficient here. This refers to the internals only and has nothing to do with how dependencies work. My intent was to have a `cargo lint` command or similar that would load and execute all lints, though.

**Reply by jhpratt** [reply]
Is there any objection to this?

**Reply by flip1995** [reply]
No objections from me. I think we got to a reasonable agreement related to this in #6

**Reply by xFrednet** [reply]
This looks good to me.

However, I'm still catching up with the discussion in #6. Could we wait on this for one more day? :upside_down_face:

**Reply by jhpratt** [reply]
Not an issue! I was just trying to "close" the uncontroversial ones by hiding them and including them in the markdown documents. Take your time, it's not like this is what's going to be holding up the whole thing 😆

**Reply by xFrednet** [reply]
I finally read through the entire discussion in #6 and this sounds good to me. +1 for adding it to the Markdown docs :upside_down_face: 

(Thank you for waiting)

**Reply by jhpratt** [reply]
Perfect, just added it. No worries about the wait.

**Comment by jhpratt** [comment]
Lints must have access to type information.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
Lints must be able to be deny, warn, and allow-by-default.

**Reply by chorman0773** [reply]
error needs to be split between deny and forbid.

**Reply by jhpratt** [reply]
Yeah, I meant deny. For lints, though, forbid-by-default doesn't make much sense, given that you could just not run them at all. It's only compiler-provided lints that could have that (and ironically don't).

**Reply by flip1995** [reply]
Forbid also makes sense for custom lints. Also its simpler to just support all lint levels. Less confusing for the consumer of the API and for the user of the linter.

Forbid is the same as Deny, just that you can't allow those lint at other points in the crate.

**Reply by jhpratt** [reply]
My point is that forbid-by-default doesn't even exist for built-in lints! I would have never guessed that except I tried it the other day. Having an equivalent of `#[forbid(foo)]` would still exist, it just couldn't be default (because you could implicitly ignore it by just not including the lint implementation).

**Reply by flip1995** [reply]
Yes, but you could make lints forbid-by-default in Clippy and in rustc. I don't see a reason why we should _forbid_ this for consumers of this API. The ones who write the lint can make this decision, no need for us to make it for them.

**Reply by jhpratt** [reply]
> you could make lints forbid-by-default in … rustc

That's not my experience when I tried it just the other day. The compiler rejected it.

How does permitting something to be forbid-by-default gain anything? If the user wants to silence it, they'll just not load that given lint, along with others that may exist in the same dylib.

**Reply by chorman0773** [reply]
The user may not *want* to silence it, but may try to do so by accident.

**Reply by jhpratt** [reply]
How would you do that on accident? It would require an explicit annotation somewhere.

**Reply by flip1995** [reply]
> The compiler rejected it.

Wait, really? I just tried changing a Clippy lint to forbid-by-default and it just worked, no complains from rustc. 

> they'll just not load that given lint

You want each lint to be loaded separately? I really think this is a bad idea. That way tools wouldn't be able to share code between lints. The consumer of the API that implements the lints should be loaded in one go and all lints should be run.

**Reply by jhpratt** [reply]
What I tried was a built-in lint for the just-landed const transmute. While the lint didn't make it in the final PR, I explicitly recall being surprised that forbid-by-default wasn't allowed in rustc. I can't speak as to clippy.

> You want each lint to be loaded separately? I really think this is a bad idea.

A given dylib could contain any number of lints. What I'm saying is that they could not load _that entire file_. When I say "load" I'm referring to the linter loading the dylib, which would be transparent to the end user.

**Reply by chorman0773** [reply]
> It would require an explicit annotation somewhere.

Or specifically you could intend to `allow` it, not knowing that there is a reason you should not. forbid by default lints are typical of things that cannot actually be done, IE. things that should not be possible to override (except with `--cap-lints`).

> Wait, really? I just tried changing a Clippy lint to forbid-by-default and it just worked, no complains from rustc.

That is... wrong. `forbid` prevents the lint from being overriden anywhere. If a lint is forbidden at some scope, it cannot be allowed, warned, or denied (but can repetatively be forbidden) within that scope.

**Reply by jhpratt** [reply]
This may be because I'm experienced with Rust, but if I'm going to `allow` a deny-by-default lint, I'm going to be certain I'm doing it for the right reason.

**Reply by flip1995** [reply]
> What I'm saying is that they could not load that entire file

IMO we shouldn't introduce another way of controlling lints. If a crate uses this API and implements lints with it, than it should be able to expect that all of those lints are run and can only be silenced with `allow` attributes and in no other way.

> That is... wrong. `forbid` prevents the lint from being overriden anywhere

Yes, I know what forbid does. I'm talking about that it is possible to make a lint forbid-by-default in Clippy and, as I just tried, in rustc. Nothing prevents you from it. Whether it makes sense is another question, but not for us to decide IMO

**Reply by chorman0773** [reply]
There are some things that you cannot do. There is no valid way to produce a null reference, or a zero NonNull. Frankly, things like ref_to_ref_mut should be forbid because that's another convoluted way of writing `core::hint::unreachable_unchecked()`.

**Reply by chorman0773** [reply]
Especially if lints are handled as regular dependencies, rather than special binaries, you wouldn't be able to easily block a forbid-by-default lint (assuming forbid-by-default worked properly). I would really like the ability to write lints for code segments that are just aliases of `unreachable_unchecked()`.

**Reply by jhpratt** [reply]
I've just opened #6 for how lints will be provided/discovered. I think this is the source of confusion and would resolve a few of the disagreements.

**Reply by jhpratt** [reply]
Given that #6 had tended towards _not_ auto-discovering lint packages, I have no particular objection to forbid-by-default lints. I presume no one else has objections to this?

**Reply by xFrednet** [reply]
I'm fine with this :+1:

**Comment by xFrednet** [comment]
Lints must be easily testable.
- for instance via UI tests which both Clippy and Rust use

**Reply by jhpratt** [reply]
compiletest-rs and trybuild already exist, which would cover this

**Reply by xFrednet** [reply]
They should cover the basics, but for the future it would definitely be better if we provide some testing support from the driver itself.

Clippy for instance uses `compiletest_rs` but still needs a bunch of additional code in the [`compile-test.rs`](https://github.com/rust-lang/rust-clippy/blob/master/tests/compile-test.rs) file to: run Clippy instead of rustc, bind to the compiler library, pass unstable feature flags to the compiler and select specific tests to be run. This manual binding to the compiler goes against the vision of this project, IMO. It can additionally lead to an ecosystem where each lint has a slightly different testing structure.

---

Another interesting test case might be to support running the lint on its own code base. I'm not sure if this is totally necessary and if it would even require additional handling. (See Clippy's [dog food test](https://github.com/rust-lang/rust-clippy/blob/master/tests/dogfood.rs))

**Reply by flip1995** [reply]
> Another interesting test case might be to support running the lint on its own code base

I don't think this makes sense for most of crate specific lints. Those lints are written for the _usage_ of the crate, which is a totally different environment than the _implementation_ of the crate.

**Reply by chorman0773** [reply]
I think having the ability makes sense, though. Not necessarily requiring it, but making it easy.

**Reply by jhpratt** [reply]
@chorman0773 In which case this should be under "could have". I don't think it's strictly necessary (or even a priority) given that _some_ tools already exist.

**Reply by flip1995** [reply]
I mean, we just get this for free, don't we? The lint package authors can just run their lints on their crate.

**Reply by jhpratt** [reply]
By "testable" I presume that it meant _automatically_ testable. Hence my response regarding compiletest-rs and trybuild.

**Reply by flip1995** [reply]
Oh wasn't sure if we talked about the testable aspect or running lints on the crate they're implemented for.

The situation with compiletest-rs is not optimal to put it lightly, so making writing tests easy is a good requirement to have. But I agree that this isn't something we have to start with right away.

**Reply by jhpratt** [reply]
Let's put this under "could have" and call it a day?

**Reply by flip1995** [reply]
Fine with me 👍
